datafusion/src/datasource/object_store/mod.rs

```diff
@@ -159,6 +159,16 @@ pub trait ObjectStore: Sync + Send + Debug {
     /// Get object reader for one file
     fn file_reader(&self, file: SizedFile) -> Result<Arc<dyn ObjectReader>>;
 }
+/// Checks if we should filter out this path name
+/// https://github.com/apache/spark/blob/5b91381b9bcea27b7be6d9cef852efbd6c23d98a/core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala#L358
+fn do_path_filter_with_suffix(path: &str, suffix: &str) -> bool {
+    let exclude = path.starts_with('_') && !path.contains('=')
+        || path.starts_with('.')
+        || path.ends_with("._COPYING_");
+    let include = path.starts_with("_common_metadata") || path.starts_with("_metadata");
+    let is_suffix = path.ends_with(suffix);
+    (!exclude || include) && is_suffix
```

Comment:
i recommend changing up the evaluation order so we can perform early exit and avoid unnecessary string operations. For example:

```rust
if !path.ends_with(suffix) {
    return false;
}

let exclude = ....;
if exlude {
    let include = path.starts_with("_common_metadata") || path.starts_with("_metadata");
    if !include {
        return false;
    }
} 

return true;
```